### ISSUE TITLE
Response payload structure

### ISSUE THREAD

**Issue by kbariotis** [octokit/octokit.js#497]
I am using the `getForRepo` function from the `issues` endpoint.

I've noticed that in some cases I got a response that it's an array with a `meta` property that contains information of the response (headers, http status, etc).

But in some other cases, the response is an object with an empty string `data` property and the `meta` property.

Why is that? Is it by design? I wasn't able to find any tests for it. 

Thank you

**Comment by kaizensoze** [comment]
Not that I know of. Can you provide an example of one endpoint for each just so I can double check?

**Comment by kbariotis** [comment]
```js
const GitHubApi = require('github');
const assert = require('assert');

const github = new GitHubApi({
  debug: false,
  protocol: 'https'
});

github.issues.getForRepo({
  owner: 'kbariotis',
  repo: 'throw.js',
  labels: 'duplicate',
  direction: 'desc'
})
.then(results => {

  assert(results.length !== undefined);
  assert(results.meta !== undefined)

  /**
   * console.log(results)
   *
   * [ meta: { 'x-ratelimit-limit': '60',
   *   'x-ratelimit-remaining': '30',
   *   'x-ratelimit-reset': '1485897767',
   *   'x-github-request-id': 'FB20:7EEC:A507490:CC439D7:5890F90E',
   *   etag: '"aacc6f617e4d9cd3fbca87b8d3748d98"',
   *   status: '200 OK' } ]
   **/

  return github.issues.getForRepo({
    headers: {
      'If-None-Match': `${results.meta.etag}`
    },
    owner: 'kbariotis',
    repo: 'throw.js',
    labels: 'duplicate',
    direction: 'desc'
  })
})
.then(results => {

  assert(results.length === undefined);
  assert(results.meta !== undefined)
  assert(results.data !== undefined)
  /**
   * console.log(results)
   *
   * { data: '',
   * meta:
   * { 'x-ratelimit-limit': '60',
   *   'x-ratelimit-remaining': '32',
   *  'x-ratelimit-reset': '1485897767',
   *   'x-github-request-id': 'FAE3:7EF2:6C5FA03:856FAFE:5890F8A4',
   *   etag: '"aacc6f617e4d9cd3fbca87b8d3748d98"',
   *   status: '304 Not Modified' } }
   **/
})
```

**Comment by kaizensoze** [comment]
Oh, honestly I have no idea. Definitely not intended by design.

**Comment by shinzui** [comment]
Are you planning to change that behavior?

**Comment by jeffwilcox** [comment]
I know I try and make use of the etag information so that in a future request I can use an `If-None-Match` header to not use a GitHub API limit call.

**Comment by kbariotis** [comment]
@jeffwilcox yeap that's im doing too. (btw have you noticed any weird behavior when using the If-None-Match header, like not always fetching the latest information?)

**Comment by jeffwilcox** [comment]
@kbariotis hmm, I haven't noticed any issues here. Are you performing paging or changing tokens? I think the etag is related to the token you use to request.

Feel free to reach out, my e-mail is jwilcox at microsoft.

**Comment by shinzui** [comment]
My issue is slightly different, maybe I should open a different issue for it. The result does not have the data keyed by `data`. 

Here is a sample code to reproduce using `8.2.1`.

```javascript
const GitHubApi = require('github')

const github = new GitHubApi({
  debug: true
})

github.repos.getForOrg({
  org: 'facebook',
  type: 'public'
}).then(repos => console.log(Object.keys(repos)))
```


```
REQUEST:  { host: 'api.github.com',
  port: 443,
  path: '/orgs/facebook/repos?type=public',
  method: 'get',
  headers: 
   { host: 'api.github.com',
     'content-length': '0',
     'user-agent': 'NodeJS HTTP Client',
     accept: 'application/vnd.github.v3+json' },
  ca: undefined }
STATUS: 200
HEADERS: {"server":"GitHub.com","date":"Fri, 10 Feb 2017 16:39:22 GMT","content-type":"application/json; charset=utf-8","content-length":"141435","connection":"close","status":"200 OK","x-ratelimit-limit":"60","x-ratelimit-remaining":"56","x-ratelimit-reset":"1486748305","cache-control":"public, max-age=60, s-maxage=60","vary":"Accept, Accept-Encoding","etag":"\"ba4775ab230e3692c553fb988469925f\"","x-github-media-type":"github.v3; format=json","link":"<https://api.github.com/organizations/69631/repos?type=public&page=2>; rel=\"next\", <https://api.github.com/organizations/69631/repos?type=public&page=6>; rel=\"last\"","access-control-expose-headers":"ETag, Link, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval","access-control-allow-origin":"*","content-security-policy":"default-src 'none'","strict-transport-security":"max-age=31536000; includeSubdomains; preload","x-content-type-options":"nosniff","x-frame-options":"deny","x-xss-protection":"1; mode=block","x-served-by":"76d9828c7e4f1d910f7ba069e90ce976","x-github-request-id":"8796:5521:210B7CC:29921CD:589DECBA"}
[ '0',
  '1',
  '2',
  '3',
  '4',
  '5',
  '6',
  '7',
  '8',
  '9',
  '10',
  '11',
  '12',
  '13',
  '14',
  '15',
  '16',
  '17',
  '18',
  '19',
  '20',
  '21',
  '22',
  '23',
  '24',
  '25',
  '26',
  '27',
  '28',
  '29',
  'meta' ]
```

**Comment by codydaig** [comment]
I'm having the same issue where it's only returning the meta, not any data.

**Comment by kaizensoze** [comment]
I think both issues that @kbariotis and @shinzui are having are a result of the following code:

https://github.com/mikedeboer/node-github/blob/master/lib/index.js#L854-L859

Some background behind the relevant change to line 855
(`git show 94e96097b1fa91d9a13a610cd3ef60f2d6f8209d`):

> commit 94e96097b1fa91d9a13a610cd3ef60f2d6f8209d
> Author: David Farr <dfarr219@gmail.com>
> Date:   Wed Jul 27 15:13:00 2016 -0700
> 
>     Avoid parsing non json data
> 
>     At least two API calls (markdown and markdown raw), return data that is not JSON.
>     We can check the content-type header and avoid parsing this data.

**Comment by kbariotis** [comment]
@kaizensoze ok the problem I see here is that the [author supposes](https://github.com/mikedeboer/node-github/blob/master/lib/index.js#L870) that the `ret` variable will always be an object, although it could also be an array.

I would suggest refactoring the above and make the output of this library be always an objects that contains both the `meta` and the `data` attributes.

What do you think?

**Comment by kaizensoze** [comment]
@kabriotis Yeah that's what I'm thinking. Always returning an object with meta and data attributes and restricting JSON.parse to the content type being application/json.

I'll need to look into any potential gotchas with this change.

**Comment by kbariotis** [comment]
@kaizensoze I would also suggest testing against various GitHub responses. Their documentation states almost all of them so It would be easy to get them as fixtures inside the test suite.

Let me know if you would like any help.

**Comment by kaizensoze** [comment]
The tests have been broken since before I took up the project to maintain it, but yeah I'll take a look.

**Comment by kaizensoze** [comment]
I've created a PR that fixes both issues mentioned. @shinzui I moved your issue into a separate ticket (https://github.com/mikedeboer/node-github/issues/504).

As for the data being an empty string in the case of providing an `If-None-Match` header, it's because of this: https://github.com/mikedeboer/node-github/blob/master/lib/index.js#L765-L779
The data variable defaults to "" and no data is received so it remains an empty string.

Also, at the risk of sounding superfluous [but looking to be helpful], @kbariotis the default params for the github api object are `debug: false, protocol: https` so you don't necessarily have to specify them.
